### PR TITLE
Enhancement/reuse session

### PR DESCRIPTION
**Pull request by iMicknl**
- Reuse session
- Abstract login method
- Add first exception logic

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -7,10 +7,14 @@ license = "MIT"
...
 repository = "https://github.com/iMicknl/python-tahoma-api"
+packages = [
+    { include = "tahoma_api" },
+    { include = "tahoma_api/**/*.py" },
+]
 
 [tool.poetry.dependencies]
 python = ">=3.6"
-aiohttp = "^3.6.2"
+aiohttp = "3.6.1"
```
Comment (tetienne) [inline pyproject.toml:17]: Why downgrade the version?

File: tahoma_api/client.py [hunk 2]
```diff
@@ -24,75 +26,55 @@ def __init__(self, username: str, password: str, api_url: str = API_URL) -> None
...
-                            "Your setup cannot be accessed through this application"
-                            in result["error"]
-                        ):
-                            print(result["error"])
-
-                        if "Bad credentials" in result["error"]:
-                            print(result["error"])
-
-                        print(result["error"])
+        self.session = aiohttp.ClientSession(raise_for_status=False)
```
Comment (tetienne) [inline tahoma_api/client.py:32]: raise_for_status is False by default: https://docs.aiohttp.org/en/stable/client_reference.html#aiohttp.request

File: tahoma_api/client.py [hunk 3]
```diff
@@ -115,53 +97,98 @@ async def fetch_event_listener(self, listener_id: str) -> List[Any]:
...
+        return response
+
+    async def get_current_executions(self) -> List[Any]:
+        """ Get all action groups executions currently running """
+        response = await self.__do_http_request("GET", "/exec/current")
+        # TODO Strongly type executions
+
+        return response
+
+    async def __do_http_request(
```
Comment (tetienne) [inline tahoma_api/client.py:142]: In a future PR, it would be nice to create dedicated get and post methods.

File: tahoma_api/client.py [hunk 4]
```diff
@@ -115,53 +97,98 @@ async def fetch_event_listener(self, listener_id: str) -> List[Any]:
...
+                result = await response.json()
 
-            if method == "POST":
-                async with session.post(
-                    self.api_url + endpoint, cookies=cookies, data=payload
-                ) as response:
-                    result = await response.json()
+        if method == "POST":
+            async with self.session.post(
+                self.api_url + endpoint, data=payload
```
Comment (tetienne) [inline tahoma_api/client.py:158]: You can use a fstring.

File: tahoma_api/client.py [hunk 5]
```diff
@@ -115,53 +97,98 @@ async def fetch_event_listener(self, listener_id: str) -> List[Any]:
...
             raise Exception
 
-        async with aiohttp.ClientSession() as session:
-            if method == "GET":
-                async with session.get(
-                    self.api_url + endpoint, cookies=cookies
-                ) as response:
-                    result = await response.json()
+        if method == "GET":
+            async with self.session.get(self.api_url + endpoint) as response:
```
Comment (tetienne) [inline tahoma_api/client.py:153]: You can use a fstring.

File: tahoma_api/models.py [hunk 6]
```diff
@@ -30,7 +31,7 @@ def __init__(
         label: str,
         deviceurl: str,
         controllable_name: str,
-        # definition: Dict[List[Any]],
+        definition: Any,
```
Comment (tetienne) [inline tahoma_api/models.py:34]: You can set the type to Dict[str, Any]

File: pyproject.toml (hunk 1 above)
Comment (iMicknl) [reply pyproject.toml:17]: Currently I am loading the integration in ha-tahoma via manifest.json. In development, this conflicts with the built-in version of aihttp in HA. I downgraded it temporarily for development.

```
  "requirements": [
    "git+https://github.com/iMicknl/python-tahoma-api.git@enhancement/reuse_session#tahoma_api===0.1.0"
  ],
```

File: tahoma_api/client.py (hunk 3 above)
Comment (iMicknl) [reply tahoma_api/client.py:142]: @tetienne agree, I am not a fan of this huge method yet.

**Pull request merged**